### PR TITLE
Improved ItemLists

### PR DESCRIPTION
**Pull request by gradientvera**

### CODE REVIEW

File: Robust.Client/UserInterface/Controls/ItemList.cs [hunk 1]
```diff
@@ -62,143 +69,146 @@ private void RecalculateContentHeight()
             _updateScrollbarVisibility();
         }
 
-        public void AddItem(string text, Texture icon = null, bool selectable = true)
+        public void Add(Item item)
```
Comment (PJB3005) [inline Robust.Client/UserInterface/Controls/ItemList.cs:72]: These `Item` instances should be made by the `ItemList` itself. With the current implementation you can theoretically add the same item to multiple lists at once, which will absolutely go haywire.

So more something like this:
```cs
public Item AddItem(string text, Texture icon = null, bool selectable = true)
```

File: Robust.Client/UserInterface/Controls/ItemList.cs [hunk 2]
```diff
@@ -548,7 +548,58 @@ public enum ItemListSelectMode
...
+
+    public sealed class Item
+    {
+        public event Action<Item> OnSelected;
+        public event Action<Item> OnDeselected;
+
+        private bool _selected = false;
+        private bool _disabled = false;
+
+        public string Text;
```
Comment (PJB3005) [inline Robust.Client/UserInterface/Controls/ItemList.cs]: All of these should be properties.

**Review by PJB3005 — CHANGES_REQUESTED** [review]

File: Robust.Client/UserInterface/Controls/ItemList.cs (hunk 2 above)
Comment (Zumorica) [reply Robust.Client/UserInterface/Controls/ItemList.cs]: Done.

File: Robust.Client/UserInterface/Controls/ItemList.cs [hunk 3]
```diff
@@ -69,7 +69,7 @@ private void RecalculateContentHeight()
             _updateScrollbarVisibility();
         }
 
-        public void Add(Item item)
+        void ICollection<Item>.Add(Item item)
```
Comment (PJB3005) [inline Robust.Client/UserInterface/Controls/ItemList.cs]: This doesn't actually prevent the item from being added by external code.

Comment (gradientvera) [reply Robust.Client/UserInterface/Controls/ItemList.cs]: Hmm, should I make Add do nothing, or just outright remove the IList<Item> interface from ItemList?

File: Robust.Client/UserInterface/Controls/ItemList.cs [hunk 4]
```diff
@@ -548,7 +565,63 @@ public enum ItemListSelectMode
         {
             None,
             Single,
-            Multiple
+            Multiple,
+            Button,
+        }
+    }
+
+    public sealed class Item
```
Comment (PJB3005) [inline Robust.Client/UserInterface/Controls/ItemList.cs]: Keep this as a nested class on `ItemList`

Comment (gradientvera) [reply Robust.Client/UserInterface/Controls/ItemList.cs]: Done! As I've mentioned in a comment, apparently [indexers compile into a property called "Item" by default](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/indexers/using-indexers#remarks) so I had to use an attribute there.

**Review by PJB3005 — APPROVED** [review]

**Pull request merged**